tests/fft_test.py

```diff
@@ -54,43 +56,83 @@ def _get_fftn_test_axes(shape):
+  # well defined derivative on the whole domain of the function, we zero-out
+  # the imaginary part of the first and possibly the last elements.
+  def wrapper(z, axes):
+    return irfft_fun(_zero_for_irfft(z, axes), axes=axes)
+  return wrapper
+
+
+def _zero_for_irfft(z, axes):
+  if axes is not None and not axes:
+    return z
+  axis = z.ndim - 1 if axes is None else axes[-1]
+  try:
+    size = z.shape[axis]
+  except IndexError:
+    return z  # only if axis is invalid, as occurs in some tests
+  if size % 2:
+    parts = [lax.slice_in_dim(z.real, 0, 1, axis=axis).real,
+             lax.slice_in_dim(z.real, 1, size - 1, axis=axis),
+             lax.slice_in_dim(z.real, size - 1, size, axis=axis).real]
+  else:
+    parts = [lax.slice_in_dim(z.real, 0, 1, axis=axis).real,
+             lax.slice_in_dim(z.real, 1, size, axis=axis)]
+  return np.concatenate(parts, axis=axis)
+
+
 class FftTest(jtu.JaxTestCase):
 
   @parameterized.named_parameters(jtu.cases_from_list(
-      {"testcase_name": "_inverse={}_shape={}_axes={}".format(
-          inverse, jtu.format_shape_dtype_string(shape, dtype), axes),
-       "axes": axes, "shape": shape, "dtype": dtype, "rng_factory": rng_factory,
-       "inverse": inverse}
+      {"testcase_name": "_inverse={}_real={}_shape={}_axes={}".format(
+          inverse, real, jtu.format_shape_dtype_string(shape, dtype), axes),
+       "axes": axes, "shape": shape, "dtype": dtype,
+       "rng_factory": rng_factory, "inverse": inverse, "real": real}
       for inverse in [False, True]
+      for real in [False, True]
       for rng_factory in [jtu.rand_default]
-      for dtype in all_dtypes
-      for shape in [(10,), (10, 10), (2, 3, 4), (2, 3, 4, 5)]
+      for dtype in (real_dtypes if real and not inverse else all_dtypes)
+      for shape in [(10,), (10, 10), (9,), (2, 3, 4), (2, 3, 4, 5)]
       for axes in _get_fftn_test_axes(shape)))
-  def testFftn(self, inverse, shape, dtype, axes, rng_factory):
+  def testFftn(self, inverse, real, shape, dtype, axes, rng_factory):
     rng = rng_factory()
     args_maker = lambda: (rng(shape, dtype),)
-    np_op = np.fft.ifftn if inverse else np.fft.fftn
-    onp_op = onp.fft.ifftn if inverse else onp.fft.fftn
+    np_op = _get_fftn_func(np.fft, inverse, real)
+    onp_op = _get_fftn_func(onp.fft, inverse, real)
     np_fn = lambda a: np_op(a, axes=axes)
-    onp_fn = lambda a: onp_op(a, axes=axes)
-    # Numpy promotes to complex128 aggressively.
+    onp_fn = lambda a: onp_op(a, axes=axes) if axes is None or axes else a
+    # Numpy promotes to complex128 aggressively.  
     self._CheckAgainstNumpy(onp_fn, np_fn, args_maker, check_dtypes=False,
                             tol=1e-4)
     self._CompileAndCheck(np_fn, args_maker, check_dtypes=True)
     # Test gradient for differentiable types.
-    if dtype in inexact_dtypes:
-      tol = 0.15  # TODO(skye): can we be more precise?
-      jtu.check_grads(np_fn, args_maker(), order=1, atol=tol, rtol=tol)
```

Comment:
Why get rid of order=1?

Reply:
`order=2` already tests all derivatives _up to_ `order=2`. So the separate `order=1` was just redundant.